lizard_auth_server/backends.py

```diff
@@ -0,0 +1,127 @@
+"""Custom Django authentication backend
+
+Copyright note: copied almost verbatim from backend.py in
+https://github.com/metametricsinc/django-warrant
+(BSD licensed)
+
+"""
+from boto3.exceptions import Boto3Error
+from botocore.exceptions import ClientError
+from django.conf import settings
+from django.contrib.auth import get_user_model
+from django.contrib.auth.backends import ModelBackend
+from django.utils.six import iteritems
+from warrant import Cognito
+
+import abc
+
+
+def cognito_to_dict(attr_list, mapping):
+    user_attrs = dict()
+    for i in attr_list:
+        name = mapping.get(i.get("Name"))
+        if name:
+            value = i.get("Value")
+            user_attrs[name] = value
+    return user_attrs
+
+
+class CognitoUser(Cognito):
+    user_class = get_user_model()
+    # Mapping of Cognito User attribute name to Django User attribute name
+    COGNITO_ATTR_MAPPING = getattr(
+        settings,
+        "COGNITO_ATTR_MAPPING",
+        {
+            "email": "email",
+            "given_name": "first_name",
+            "family_name": "last_name",
+        },
+    )
+
+    def get_user_obj(self, username=None, attribute_list=[], metadata={}, attr_map={}):
+        user_attrs = cognito_to_dict(attribute_list, CognitoUser.COGNITO_ATTR_MAPPING)
+        django_fields = [f.name for f in CognitoUser.user_class._meta.get_fields()]
+        extra_attrs = {}
+        for k, v in user_attrs.items():
+            if k not in django_fields:
+                extra_attrs.update({k: user_attrs.pop(k, None)})
+        if getattr(settings, "COGNITO_CREATE_UNKNOWN_USERS", True):
+            user, created = CognitoUser.user_class.objects.update_or_create(
+                username=username, defaults=user_attrs
```

Comment:
Zoals ik het begrijp is het creëren van een lokale (SSO) user nodig om een django session te kunnen aanmaken die voor de permanente login in de SSO zorgt. Dus inderdaad moet COGNITO_CREATE_UNKNOWN_USERS op True.

Wel ben ik bang dat dit leidt tot vreemde situaties.

- Ik geloof (moet getest worden) dat er via een trigger een UserProfile wordt aangemaakt (https://github.com/lizardsystem/lizard-auth-server/blob/master/lizard_auth_server/signal_handlers.py#L10). Mocht dat niet zo zijn: SSO verwacht op sommige punten dat UserProfile bestaat.
- Users die wél bestaan in Cognito, maar niet in de SSO die worden hier aangemaakt. We moeten ervoor zorgen dat ze nooit weer naar Cognito gemigreerd kunnen worden. Dus ik denk: UserProfile bijwerken hier met `migrated_at = django.utils.timezone.now()` in geval `created = True`.
- Users die gemigreerd zijn uit de SSO krijgen hier een update van hun metadata!? Is denk ik geen probleem, maar lijkt me niet nodig. Kost een SQL UPDATE voor iedere authenticatie. Een `get_or_create` lijkt me meer op zijn plek.

Reply:
- get_or_create gebruikt.
- alleen bij create wordt migrated_at gezet (en voor de volledigheid ook maar de andere attributen, dat kan geen kwaad).
- Userprofile gaat altijd automatisch.